Adds FORCE_MULTILINE_DICT knob

brianmego opened the pull request:
I often find myself wanting to format my dictionaries with keys on their own lines, ala:

```
my_dict = {
    'spam': 'eggs',
    'foo': 'bar'
}
```

EACH_DICT_ENTRY_ON_SEPARATE_LINE did previous exist, but would only fire if the length of the dictionary was greater than the max column length. This patch allows it to always be turned on with the FORCE_MULTILINE_DICT config option, even for short dictionaries.

- Defaults to False (off)
- Adds docs
- Adds tests
- While in there, fixed two flaky tests that were dependent on previous tests setting up the environment

NOTE: There may be a better way to do that LBRACE check in reformatter.py. A quick glance through the codebase didn't show anything obvious, but let me know if I'm redoing work here.

brianmego commented:
Any thoughts on this?

brianmego commented on yapftests/reformatter_basic_test.py:
```diff
@@ -1545,11 +1545,13 @@ def f(self, aaaaaaaaa, bbbbbbbbbbbbb, row):
     self.assertCodeEqual(code, reformatter.Reformat(uwlines))
 
   def testFunctionCallInDict(self):
+    style.SetGlobalStyle(style.CreateChromiumStyle())
```
You're not wrong, but also, if I leave this out:

```
FAIL: testFunctionCallInDict (yapftests.reformatter_basic_test.BasicReformatterTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/brian/Code/yapf/yapftests/reformatter_basic_test.py", line 1551, in testFunctionCallInDict
    self.assertCodeEqual(code, reformatter.Reformat(uwlines))
  File "/home/brian/Code/yapf/yapftests/yapf_test_helper.py", line 57, in assertCodeEqual
    self.fail('\n'.join(msg))
AssertionError: Code format mismatch:
Expected:
 > a = {'a': b(c=d, **e)}
Actual:
 > a = {
 >     'a': b(c=d, **e)
 > }
Diff:
--- actual
+++ expected
@@ -1,3 +1 @@
-a = {
-    'a': b(c=d, **e)
-}
+a = {'a': b(c=d, **e)}
```

There's some flakiness in the test suite somewhere regarding a static reference, but rather than tracking that down, I just took this 'ensure our environment is what we think it is' approach.

gwelymernans replied:
Ah! I see what's going on. When you modify the style in the tests you added, you need to do that within a try-except block. There should be some examples in the yapftests directory.

Also make sure to run `python -m yapf -r -i .` to make sure it passes the formatting checks.

brianmego replied:
Rock on. Took care of it. Is HACKING.rst a standard I've just never seen before? That's where the "run yapf on itself to verify integrity" docs lived. Didn't even think to look there!

gwelymernans replied:
The HACKING.rst is pretty bare bones as it is, but should be generally useful. It has information on how to release new versions that I follow. :-)

brianmego commented:
Thanks for the review! Spacing/README/CHANGELOG/paren patches have all been applied. Regarding the test note, I cannot remove that and also have the tests pass. Perhaps the addition of the new tests exposed a bit of flakiness in environment?

brianmego commented:
Am I taking crazy pills with this test coverage? As far as I can tell, every line I added is now tested. I was so frustrated trying to find the line that decreased coverage that I just found some untested code and added tests around it, that made the coverage go down MORE! What am I missing here?

brianmego commented:
Ok, finally got it through my thick head how to read those jenkins messages, apparently. All good from my end.

brianmego commented:
Bump. Anything else you're looking for in this branch?
